### PR TITLE
Add implementation of discovery API to Custom Metrics - Stackdriver Adapter

### PR DESCRIPTION
**Pull request by kawych**

### CODE REVIEW

**Comment by kawych** [comment]
@DirectXMan12 
Please take a look.

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go [hunk 1]
```diff
@@ -48,24 +51,20 @@ type Translator struct {
...
 	if len(resourceIDs) == 0 {
-		return nil, fmt.Errorf("No objects matching provided selector")
+		return nil, apierr.NewBadRequest("No objects matched to provided selector")
 	}
-	filter, err := t.createFilter(resourceIDs, t.config.Project, t.config.Zone, t.config.Cluster, metricName)
-	if err != nil {
-		return nil, err
-	}
-	return t.createListTimeseriesRequest(filter), nil
+	return t.createListTimeseriesRequest(joinFilters(t.filterForMetric(t.config.MetricsPrefix+"/"+metricName), t.filterForCluster(), t.filterForPods(resourceIDs))), nil
```
Comment (piosz) [inline custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go]: it's unreadable
could you please extract some params to local vars?

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go [hunk 2]
```diff
@@ -76,19 +75,46 @@ func (t *Translator) GetRespForPod(response *stackdriver.ListTimeSeriesResponse,
...
+	values, err := t.getMetricValuesFromResponse(groupResource, namespace, response, metricName)
 	if err != nil {
 		return nil, err
 	}
 	return t.metricsFor(values, groupResource, metricName, podList)
 }
 
+// ListMetricDescriptors returns Stackdriver request for all custom metrics descriptors.
+func (t *Translator) ListMetricDescriptors() *stackdriver.ProjectsMetricDescriptorsListCall {
+	return t.service.Projects.MetricDescriptors.List(fmt.Sprintf("projects/%s", t.config.Project)).
```
Comment (piosz) [inline custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go]: same here

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go [hunk 3]
```diff
@@ -76,19 +75,46 @@ func (t *Translator) GetRespForPod(response *stackdriver.ListTimeSeriesResponse,
...
+		Filter(joinFilters(t.filterForMetricPrefix(), t.filterForCluster(), t.filterForAnyObject()))
+}
+
+// GetMetricsFromSDDescriptorsResp returns an array of MetricInfo for all metric descriptors
+// returned by Stackdriver API that satisfy the requirements:
+// - metricKind is "GAUGE"
+// - valueType is "INT64" or "DOUBLE"
```
Comment (piosz) [inline custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go]: Do we support both in the API?

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go [hunk 4]
```diff
@@ -97,24 +123,37 @@ func getResourceIDs(list *v1.PodList) []string {
...
-	zoneFilter := fmt.Sprintf("resource.label.zone = %q", zone)
-	clusterFilter := fmt.Sprintf("resource.label.cluster_name = %q", strings.TrimSpace(clusterName))
-	// container_name is set to empty string for pod metrics
+func joinFilters(filters ...string) string {
+	return strings.Join(filters, " AND ")
+}
+
+func (t *Translator) filterForCluster() string {
+	projectFilter := fmt.Sprintf("resource.label.project_id = %q", t.config.Project)
+	zoneFilter := fmt.Sprintf("resource.label.zone = %q", t.config.Zone)
```
Comment (piosz) [inline custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go]: What if we run nodes in multiple zones? Could you please add TODO and create a bug entry to figure it out.

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go (hunk 1 above)
Comment (kawych) [reply custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go]: Done.

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go (hunk 2 above)
Comment (kawych) [reply custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go]: Done.

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go (hunk 3 above)
Comment (kawych) [reply custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go]: Custom Metrics API uses `resource.Quantity` and doesn't say anything about underlying SD API. I think it's up to us what we support and the most obvious decision seems to be to allow any numeric value.

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go (hunk 4 above)
Comment (kawych) [reply custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go]: Does it matter? I think what matters is that (project, zone, cluster name, pod id, metric name) identifies the SD time series unambiguously. The function name may be misleading for the reason you mentioned, what about just renaming/splitting it?

Comment (piosz) [reply custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go]: Yes. Let's imagine you have pod running a node in zone1, and the adapter runs in zone2. You will query for zone2 metrics while they are attached to zone1.

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go (hunk 3 above)
Comment (piosz) [reply custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go]: SGTM

**Comment by piosz** [comment]
/lgtm
Please squash commits before merge.

File: custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go (hunk 4 above)
Comment (kawych) [reply custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go]: Right, obviously. Added a TODO.

**Pull request merged**